### ISSUE TITLE
Bidirectional many-to-many

### ISSUE THREAD

**Issue by mindplay-dk** [directus/directus#2650]
# Feature Request

I've been trying out the live demo for the following use-case:

I'm building a shop, in which products can have related products.

For example, a "Blue Shirt" is available in small, medium and large - similarly to e.g. Amazon, these will be listed as 3 separate products, and each product page will display links to products related by size.

I created a `size` field as a drop-down, and a `other_sizes` field as a many-to-many:

![image](https://user-images.githubusercontent.com/103348/78118508-66696e00-7407-11ea-8e16-96e95ac78c0d.png)

I created the products and entered the other sizes, but this only works one way:

![image](https://user-images.githubusercontent.com/103348/78118444-4cc82680-7407-11ea-80ca-0dd300eb436d.png)

What is expected for this use-case is a bidirectional relationship: the content manager shouldn't need to duplicate these relationships product by product, and instead the relationships should work in both directions.

## What problem does this feature solve?

Scenarios such as related articles, related products, friendship, and so on - any many-to-many relationship where the direction does not matter.

## How do you think this should be implemented?

This could be implemented as a checkbox or drop-down option in the Many to Many field type.

At the SQL level, there are a few different ways to store and query bidirectional relationships - at least:

1. Store low/high key tuples, e.g. `1, 2` and *not* `2, 1`.

2. Store duplicate tuples, e.g. `1, 2` and `2, 1`, and insert/delete these transactionally.

Where (2) may be slightly easier to query, twice as much storage is required for the records and indices, so I would lean towards (1) personally, e.g.:

```sql
CREATE TABLE other_product_sizes
(
    low_product_id INT NOT NULL,
    high_product_id INT NOT NULL,
    CONSTRAINT one_way CHECK (low_product_id < high_product_id),
    PRIMARY KEY (low_product_id, high_product_id)
);
```

Note that `CHECK` constraints are [new](https://mysqlserverteam.com/mysql-8-0-16-introducing-check-constraint/) to MySQL 8, though (as I recall) the syntax was supported in earlier versions for compatibility, although the constraints weren't actually checked.

Note that my example is self-referencing - it may be similar to #2374, not sure, the description/discussion was a bit hard to follow.

I've also seen #749, but that appears to be about the UI, whereas (I think) what I'm describing is about the model - if this was a separate field-type (rather than an option for the Many to Many field-type) the UI probably could be identical to that of the Many to Many field.

## Would you be willing to work on this?

I haven't decided to use your product yet, just evaluating - but this feature seemed important enough that it was worth reporting.

It also looks like you're in the middle of rewriting the entire product with proper dependency injection and tests. (Something that will definitely make the product much more attractive to someone like me! although this already looks like a pretty strong product! 🙂)

**Comment by benhaynes** [comment]
Maybe I'm misunderstanding, but are you saying that when you create a M2M from Collection A->B, you also want the same relationship available from B->A (in both the App and API responses)?

**Comment by mindplay-dk** [comment]
> Maybe I'm misunderstanding, but are you saying that when you create a M2M from Collection A->B, you also want the same relationship available from B->A (__in both the App and API responses__)?

Yep. 🙂

Technically, this creates an [undirected graph](https://mathinsight.org/definition/undirected_graph), as every product can be considered a *node*, and every entry in the link-table an *edge* - and while relational databases aren't very good at querying graphs, for use-cases such as related articles or products, we're typically only interested in the immediate neighbors of these nodes. A web page, for example, would display an article or product, and then list immediately related ones. (Use-cases requiring more than that should probably use a graph database.)

**Comment by rijkvanzanten** [comment]
You can add a many to many interface to the other collection and use the same junction table for the relationship (that's also selected by default) to set this up 🙂

**Comment by benhaynes** [comment]
And as a follow up, we're working on a way in v9 to automate this reciprocal field... see mockup below:

![Settings-Data-Model-Field-Create-2B](https://user-images.githubusercontent.com/522079/78159686-e3e5ac00-7410-11ea-96b9-2197aab181d6.png)

**Comment by mindplay-dk** [comment]
> You can add a many to many interface to the other collection and use the same junction table for the relationship (that's also selected by default) to set this up 🙂

You're describing a *directed* many-to-many relationship between *two* collections.

The use-case I described is products relating to other products - so there is no "other collection" in the type of relationship I'm trying to model here.

Let me use a very [concrete example](https://www.amazon.co.uk/SanDisk-126905-Ultra-Flair-Flash/dp/B015CH1JIW/ref=sr_1_4?dchild=1&keywords=usb+stick&qid=1585822374&sr=8-4):

![image](https://user-images.githubusercontent.com/103348/78237459-a4d05d00-74db-11ea-8442-1a573c64884a.png)

As you can see from the highlighted product title (and URL) we're on a page specifically about the 32GB size and silver/black color version of this product.

The size and color options navigate to different products - that is, this isn't just one product with a list of options, they're different products being linked together. Each product has a different product IDs, a distinct URL, different photos, description, product number, price, availability, and so on.

The desired UI for content management is precisely that of the many-to-many interface: I just want a list of products related by size, and another list of products related by color.

But the resulting model when saved/queried is something very different: if the content manager creates a product `1`, and adds products `2` and `3` as being related by size, that *implies* products `2` and `3` are also related to product `1`. The content manager shouldn't need to go to products `2` and `3` individually and input the same relationships again.

The only solution I can think of at the moment, is to create a separate collection, e.g. `product_groups`, which would have a many-to-many relationship with `products`.

That's probably the right way to *model* it - and I'm realizing now as I describe it, that this is probably something different from the friendship model, since friendships are bidirectional *1-to-1*, whereas something like related articles or products is more likely bidirectional *many-to-many*.

Anyhow, modeling it this way doesn't result in the right UI. The content manager shouldn't see "Product Groups" in the menu - they shouldn't need to leave the product page and then create two product groups as separate items in a separate collection. All the user should need to do is pick the related products and press save, just like the many-to-many UI.

I don't think there's a field type or option that leads to that experience?

**Comment by benhaynes** [comment]
Ahh, that makes more sense — and to be honest, I assumed that a M2M to the same collection would work in both directions... but perhaps we need to look into its inner workings. Perhaps this should be a feature request? @directus/team ?

**Comment by rijkvanzanten** [comment]
> Anyhow, modeling it this way doesn't result in the right UI. The content manager shouldn't see "Product Groups" in the menu - they shouldn't need to leave the product page and then create two product groups as separate items in a separate collection. All the user should need to do is pick the related products and press save, just like the many-to-many UI.

I think a setup like follows might work for that use case:

![image](https://user-images.githubusercontent.com/9141017/78268210-dcd4a180-74d5-11ea-8b47-8d47e3ea5c38.png)

A product has multiple variants, and each variant has a single size and color. You can put fields that are the same for every product on this top level product level, and add more descriptive things like specialty images and descriptions on the variants level. It's similar to the "product group" idea, but by changing the nomenclature, I believe you end up with the situation you're describing

**Comment by mindplay-dk** [comment]
@rijkvanzanten yes, that would work, but you're describing a different model that works for the UI - but that's not the model this project calls for. 🙂

> You can put fields that are the same for every product on this top level product level, and add more descriptive things like specialty images and descriptions on the variants level.

Products literally have nothing in common - so the products table would be an empty table with just an ID. The variants table would have all the product information, such as size and color, URL tokens, photos, description, product codes, price, inventory numbers, and so on.

Renaming your products table to "product groups" and your variants table to "products" would be more accurate - and then you're back to the model where the user starts by editing a product group and then has to add products to it, which is what I don't want.

Another reason this model doesn't work is, not all products have variants. So where would this information go? Either, you'd need to have all of the product schema duplicated between the products and variants tables, with variants somehow "overriding" product fields, which sounds very fragile and complex - or, every product would need to have at least "one variant", which doesn't really align with the reality I'm trying to model either.

The reality is that physical products really are unique products in every sense, and some of them happen to have a relationship where one product is a meaningful alternative option to another product. They get listed, added to a shopping-cart, bought, shipped, sold out, and so on, as unique individual items with no real relationship beyond a useful reference to other products the user might choose instead.

> The last part of this puzzle is adding a secondary m2m to the products table that relates back to products, so you can have related products

But then we're back to the original problem, right? 🙂

If that were possible, it wouldn't matter if I were building the model you described, or the model I'm describing - it would work equally well for both.

Either way, it could sound a bit like you're trying to design a model that works for the CMS? 🙂

Ideally, I don't think the CMS admin UI should be driving decisions about your domain model - that should emerge from analysis of user/technical requirements and the design of the application itself.

> to be honest, I assumed that a M2M to the same collection would work in both directions... but perhaps we need to look into its inner workings.

@benhaynes it's probably not that simple. 🙂

There are examples of both directed and undirected many-to-many relationships, so you probably don't want self-referencing many-to-many fields to work that way exclusively.

For example, "following" on Twitter is a directed many-to-many relationship: I happen to follow your account `x -> y`, but that does not imply you're following my account `y -> x`.

Counter-example, "related articles" in a CMS: if article A is related to article B, then article A gets listed as related on article B's page, *and* vice-versa - the relationship is undirected, `a <-> b`, there's no direction on that arrow.

And lastly, "related products": I might call this a many-to-many *group* relationship - there's difference between this and the "related articles" example, which doesn't become apparent until you have more than two related items, at least three.

To illustrate the difference, for "related articles", if the user adds these relationships:

    a -> b
    a -> c

This implies a reverse relationship:

    b <- a
    c <- a

Or simplifying the notation, just:

    a <-> b
    a <-> c

This does *not* imply any relationship between `b` and `c` - articles are related in "pairs of two", but that relationship has no meaning beyond the immediate relatedness of those two articles.

Now, for "related products", this is a different story - if the user adds these relationships:

    a -> b
    a -> c

This implies a *group* relationship between *all* of these nodes to a group identity:

    a -> G
    b -> G
    c -> G

What doesn't seem possible right now is configuring an appropriate UI for this?

I think the best we can do right now is ask the user to create products first, and then go to a "size groups" collection, create one, and list the products that are related by size. Not a good workflow.

What is desired here, is a means of creating and adding to that group directly from the product admin page. If it were possible to not see or create these groups explicitly at all, e.g. just present the user with a list of related products, even better.

One more thing to consider though, is might this group table also have properties of it's own? For example, in my case, I could make do with two of these group tables, one for size and one for color, since those are the only relationships required for my use-case. One could imagine larger shops having many ways in which products could be related, so you might add a drop-down for the type of relationship, at which point the UI probably looks quite different...

I feel like I'm digressing here, but it's a deep subject - I'd like to suggest a simple solution that covers all of these use-cases, but... UI is hard. 😊
